google-oauth-client/src/main/java/com/google/api/client/auth/openidconnect/IdTokenVerifier.java

```diff
@@ -280,5 +455,174 @@ public Builder setAcceptableTimeSkewSeconds(long acceptableTimeSkewSeconds) {
+    }
+
+    /** Sets the environment. Used mostly for testing */
+    Builder setEnvironment(Environment environment) {
+      this.environment = environment;
+      return this;
+    }
+
+    /**
+     * Set the HttpTransportFactory used for requesting public keys from the certificate URL. Used
+     * mostly for testing.
+     *
+     * @param httpTransportFactory the HttpTransportFactory used to build certificate URL requests
+     * @return the builder
+     */
+    public Builder setHttpTransportFactory(HttpTransportFactory httpTransportFactory) {
+      this.httpTransportFactory = httpTransportFactory;
+      return this;
+    }
+  }
+
+  /** Custom CacheLoader for mapping certificate urls to the contained public keys. */
+  static class PublicKeyLoader extends CacheLoader<String, Map<String, PublicKey>> {
+    private final HttpTransportFactory httpTransportFactory;
+
+    /**
+     * Data class used for deserializing a JSON Web Key Set (JWKS) from an external HTTP request.
+     */
+    public static class JsonWebKeySet extends GenericJson {
+      @Key public List<JsonWebKey> keys;
+    }
+
+    /** Data class used for deserializing a single JSON Web Key. */
+    public static class JsonWebKey {
+      @Key public String alg;
+
+      @Key public String crv;
+
+      @Key public String kid;
+
+      @Key public String kty;
+
+      @Key public String use;
+
+      @Key public String x;
+
+      @Key public String y;
+
+      @Key public String e;
+
+      @Key public String n;
+    }
+
+    PublicKeyLoader(HttpTransportFactory httpTransportFactory) {
+      super();
+      this.httpTransportFactory = httpTransportFactory;
+    }
+
+    @Override
+    public Map<String, PublicKey> load(String certificateUrl) throws Exception {
+      HttpTransport httpTransport = httpTransportFactory.create();
+      JsonWebKeySet jwks;
+      try {
+        HttpRequest request =
+            httpTransport
+                .createRequestFactory()
+                .buildGetRequest(new GenericUrl(certificateUrl))
+                .setParser(GsonFactory.getDefaultInstance().createJsonObjectParser());
+        HttpResponse response = request.execute();
+        jwks = response.parseAs(JsonWebKeySet.class);
+      } catch (IOException io) {
+        LOGGER.log(
+            Level.WARNING,
+            "Failed to get a certificate from certificate location " + certificateUrl,
+            io);
+        return ImmutableMap.of();
+      }
+
+      ImmutableMap.Builder<String, PublicKey> keyCacheBuilder = new ImmutableMap.Builder<>();
+      if (jwks.keys == null) {
+        // Fall back to x509 formatted specification
+        for (String keyId : jwks.keySet()) {
+          String publicKeyPem = (String) jwks.get(keyId);
+          keyCacheBuilder.put(keyId, buildPublicKey(publicKeyPem));
+        }
+      } else {
+        for (JsonWebKey key : jwks.keys) {
+          try {
+            keyCacheBuilder.put(key.kid, buildPublicKey(key));
```

Comment:
I believe `buildPublicKey` returning null would be unexpected. But, for the future writer or troubleshooter, would you add logging for such null case?

Or a source code comment explaining that buildPublicKey doesn't return null would be helpful. (I read that the null case shouldn't happen happen because the algorithm field already checked for the valid values.)